lib/spack/env/cc

```diff
@@ -573,38 +580,72 @@ while [ $# -ne 0 ]; do
+            shift
+            if [ $# -eq 0 ]; then
+                # -Xlinker without value: let the compiler error about it.
+                append other_args_list -Xlinker
+                xlinker_expect_rpath=no
+                break
+            elif [ "$xlinker_expect_rpath" = yes ]; then
+                # Register the path of -Xlinker -rpath <other args> -Xlinker <path>
+                if system_dir "$1"; then
+                    append system_rpath_dirs_list "$1"
+                else
+                    append rpath_dirs_list "$1"
                 fi
-                shift 3;
-                rp="$1"
-            elif [ "$2" = "$dtags_to_strip" ]; then
-                shift  # We want to remove explicitly this flag
+                xlinker_expect_rpath=no
             else
-                append other_args_list "$1"
+                case "$1" in
+                    -rpath=*)
+                        arg="${1#-rpath=}"
+                        if system_dir "$arg"; then
+                            append system_rpath_dirs_list "$arg"
+                        else
+                            append rpath_dirs_list "$arg"
+                        fi
+                        ;;
+                    --rpath=*)
+                        arg="${1#--rpath=}"
+                        if system_dir "$arg"; then
+                            append system_rpath_dirs_list "$arg"
+                        else
+                            append rpath_dirs_list "$arg"
+                        fi
+                        ;;
+                    -rpath|--rpath)
+                        xlinker_expect_rpath=yes
+                        ;;
+                    "$dtags_to_strip")
+                        ;;
+                    *)
+                        append other_args_list -Xlinker
+                        append other_args_list "$1"
+                        ;;
+                esac
             fi
             ;;
+        "$dtags_to_strip")
+            ;;
         *)
-            if [ "$1" = "$dtags_to_strip" ]; then
-                :  # We want to remove explicitly this flag
-            else
-                append other_args_list "$1"
-            fi
+            append other_args_list "$1"
             ;;
     esac
-
-    # test rpaths against system directories in one place.
-    if [ -n "$rp" ]; then
-        if system_dir "$rp"; then
-            append system_rpath_dirs_list "$rp"
-        else
-            append rpath_dirs_list "$rp"
-        fi
-    fi
     shift
 done
 
+# We found `-Xlinker -rpath` but no matching value `-Xlinker /path`. Just append
+# `-Xlinker -rpath` again and let the compiler or linker handle the error during arg
+# parsing.
+if [ "$xlinker_expect_rpath" = yes ]; then
+    append other_args_list -Xlinker
+    append other_args_list -rpath
+fi
+
+# Same, but for -Wl flags.
+if [ "$wl_expect_rpath" = yes ]; then
+    append other_args_list -Wl,-rpath
+fi
```

Comment:
If I get what this is doing, this only happens at the end of an arg list, right?  i.e., if we have a trailing `-Xlinker` option with no argument.  In that case, why not just leave it and only append `-Xlinker`?  Appending `-Xlinker -rpath` will obfuscate the real error -- whereas passing it through as-is will at least generate the same error.

Reply:
There's these cases:

```
-Xlinker -rpath --other --flag
-Xlinker -rpath --other --flag -Xlinker
```

we always pop `-Xlinker -rpath` expecting a value, so the compiler line becomes

```
--other --flag
--other --flag -Xlinker
```

appending the unused `-Xlinker -rpath`  again, you'd get:

```
--other --flag -Xlinker -rpath # fine, same error
--other --flag -Xlinker -Xlinker -rpath # wrong: parsed as `-Xlinker -Xlinker` `-rpath`, unrecognized `-rpath` option for the compiler, or unrecognized `-Xlinker` flag for the linker
```

So the current implementation drops `-Xlinker -rpath` when the last bit is `-Xlinker` without argument:

```
--other --flag -Xlinker -rpath
--other --flag -Xlinker # drop the -Xlinker, compiler will error with the same -Xlinker has no args error anyways.
```

Reply:
i mean, this is a terrible edge case... I think we don't really handle `-L`/`-I` etc without argument either...

Reply:
Ok I think this one's fine.